src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/net/NetClient.java

```diff
@@ -41,26 +48,39 @@ public NetClient(final GRPCConnectionManager connectionManager) {
     this.connectionManager = connectionManager;
   }
 
-  private Map<String, StreamObserver<FlowUnitMessage>> perHostOpenDataStreamMap = new HashMap<>();
+  private ConcurrentMap<String, AtomicReference<StreamObserver<FlowUnitMessage>>> perHostOpenDataStreamMap =
+      new ConcurrentHashMap<>();
 
   public void subscribe(
       final String remoteHost,
       final SubscribeMessage subscribeMessage,
       StreamObserver<SubscribeResponse> serverResponseStream) {
     LOG.debug("Trying to send intent message to {}", remoteHost);
-    connectionManager
-        .getClientStubForHost(remoteHost)
-        .subscribe(subscribeMessage, serverResponseStream);
+    try {
+      connectionManager
+          .getClientStubForHost(remoteHost)
+          .subscribe(subscribeMessage, serverResponseStream);
+    } catch (StatusRuntimeException sre) {
+      LOG.error("Encountered an error trying to subscribe. Status: {}",
+          sre.getStatus(), sre);
+      StatsCollector.instance().logException(StatExceptionCode.RCA_NETWORK_ERROR);
+    }
   }
 
   public void publish(
       final String remoteHost,
       final FlowUnitMessage flowUnitMessage,
       final StreamObserver<PublishResponse> serverResponseStream) {
     LOG.debug("Publishing {} data to {}", flowUnitMessage.getGraphNode(), remoteHost);
-    final StreamObserver<FlowUnitMessage> stream =
-        getDataStreamForHost(remoteHost, serverResponseStream);
-    stream.onNext(flowUnitMessage);
+    try {
+      final StreamObserver<FlowUnitMessage> stream =
+          getDataStreamForHost(remoteHost, serverResponseStream);
+      stream.onNext(flowUnitMessage);
+    } catch (StatusRuntimeException sre) {
+      LOG.error("Encountered an error trying to publish a flow unit. Status: {}",
```

Comment:
What happens to stale connections? When do we throw them away? Can we have connection failures where we re-establish a connection?

The case we need to test is - 

* Node joins cluster.
* You are streaming metrics.
* Node leaves cluster temporarily and then joins back.
* Will we create a new connection and stream data?

Reply:
We should also track cluster membership changes to add/remove nodes.

Reply:
Stale connections will result in subscribers resending a single subscribe request and re-establishing the stream.

Added mechanism to check and trigger subscription if the cluster membership reports nodes that the network layer has not seen. And unsubscribe/terminate connection if cluster membership says a node is not part of the cluster.